### PR TITLE
Create request and response timing and errors metric hooks

### PR DESCRIPTION
**Pull request by varunravi98**
This PR adds hooks to measure request and response time and track the number of errors encountered during this process. The hook is designed to be analogous to `ProxyInterceptor` and `ProxyInterceptorFactory`, with `MetricsHook` and `MetricsHookFactory`, respectively. Each `ProxyConfig` now contains a `MetricsHookFactory` field called `CollectorHookFactory`. A non-nil type that implements `CollectorHookFactory` implements the `NewHook()` method, which enables it to create new `MetricsHook`s. Meanwhile, the `ProxySession` type now includes two additional fields: `hooks`, which is a map of `MetricsHook`s, and `isMetricsEnabled`, a bool. When `CreateWorker` is called to generate a new `ProxySession`, `hooks` is populated with `MetricsHook`s generated from the `ProxyConfig`'s `CollectorHookFactory` field (provided that it is non-nil). At the moment, the `MetricsHook`s included are all the ones that are needed for the metrics being tracked, but more can be easily added as necessary. Meanwhile, `isMetricsEnabled` is set to true only if the `CollectorHookFactory` field was non-nil. If it was `nil`, then none of the hooks can be created, and `isMetricsEnabled` is set to false so that any hook updates are ignored. This should only occur during `mongonet`-only tests.

Each type implementing a `MetricHook` implements `StartTimer()`, `StopTimer()`, `SetGauge()`, `AddCounterGauge()`, `IncCounterGauge()`, `DecGauge()`, and `SubGauge()`. The type could wrap (for example) a Prometheus metric, and by implementing these methods, it will be able to perform any basic action on its underlying metric. Any metric hook can be accessed from a `ProxySession` via the hooks map, provided that the programmer uses the same string key to access the metric as they did to insert it into the map in `CreateWorker()`. Lastly, each hook is utilized wherever necessary, using the appropriate method to modify the underlying metric provided that the hook was initialized in the first place (hence all of the `isMetricsEnabled` checks).

### CODE REVIEW

File: proxy.go [hunk 1]
```diff
@@ -390,11 +406,43 @@ func wrapNetworkError(err error) error {
...
+		}
+		responseErrorsHook, ok = ps.hooks["responseErrorsHook"]
+		if !ok {
+			return nil, fmt.Errorf("could not access the response processing errors metric hook")
+		}
+	}
+
 	// reading message from client
+	if ps.isMetricsEnabled {
+		err := requestDurationHook.StartTimer()
```
Comment (TomerYakir) [inline proxy.go:433]: Let's start the timer after the message is successfully read so we don't count idle client time.
I think we're fine not timing failed attempts to read client messages.

File: proxy.go [hunk 2]
```diff
@@ -390,11 +406,43 @@ func wrapNetworkError(err error) error {
...
+		if err != nil {
+			ps.logTrace(ps.proxy.logger, ps.proxy.config.TraceConnPool, "failed to start request duration metric hook timer %v", err)
```
Comment (TomerYakir) [inline proxy.go:435]: `logTrace` logs at debug-level and is expected to be turned off by default. Let's change to `ps.proxy.logger.Logf(slogger.WARN, ....)`

File: proxy.go [hunk 3]
```diff
@@ -441,15 +501,25 @@ func (ps *ProxySession) doLoop(mongoConn *MongoConnectionWrapper) (*MongoConnect
...
+				requestErrorsHook.IncCounterGauge()
+			}
 			ps.logTrace(ps.proxy.logger, ps.proxy.config.TraceConnPool, "failed to get a new connection. err=%v", err)
 			return nil, NewStackErrorf("cannot get connection to mongo %v using connection mode=%v readpref=%v", err, ps.proxy.config.ConnectionMode, rp)
 		}
 		ps.logTrace(ps.proxy.logger, ps.proxy.config.TraceConnPool, "got new connection %v using connection mode=%v readpref=%v", mongoConn.conn.ID(), ps.proxy.config.ConnectionMode, rp)
 	}
 
+	if ps.isMetricsEnabled {
+		requestDurationHook.StopTimer()
```
Comment (TomerYakir) [inline proxy.go:514]: [q] how do we distinguish between request_proxy_only and request_total?

File: proxy.go [hunk 4]
```diff
@@ -467,13 +537,27 @@ func (ps *ProxySession) doLoop(mongoConn *MongoConnectionWrapper) (*MongoConnect
...
+				responseErrorsHook.IncCounterGauge()
+			}
 			mongoConn.ep.ProcessError(wrapNetworkError(err), mongoConn.conn)
 			return nil, NewStackErrorf("error reading wire message from mongo conn %v: %v", mongoConn.conn.ID(), err)
 		}
+
+		if ps.isMetricsEnabled {
+			err = responseDurationHook.StartTimer()
+			if err != nil {
+				ps.logTrace(ps.proxy.logger, ps.proxy.config.TraceConnPool, "failed to start response duration metric hook timer %v", err)
```
Comment (TomerYakir) [inline proxy.go:550]: same comment about logging

File: proxy.go [hunk 5]
```diff
@@ -482,26 +566,43 @@ func (ps *ProxySession) doLoop(mongoConn *MongoConnectionWrapper) (*MongoConnect
...
+				if ps.isMetricsEnabled {
+					responseErrorsHook.IncCounterGauge()
+				}
 				return nil, NewStackErrorf("error intercepting message %v", err)
 			}
 		}
+
 		ps.logMessageTrace(ps.proxy.logger, ps.proxy.config.TraceConnPool, resp)
+		if ps.isMetricsEnabled {
+			responseDurationHook.StopTimer()
```
Comment (TomerYakir) [inline proxy.go:596]: where do we distinguish between response_total and response_proxy_only?

**Review by TomerYakir — CHANGES_REQUESTED** [review]

**Comment by TomerYakir** [comment]
Small comments from me.
One thing to note is that `mongonet` is actually agnostic to the way metrics are actually implemented, so I'd rephrase `The type must wrap a Prometheus metric` to `The type could (for example) wrap a Prometheus metric`.

File: proxy.go [hunk 6]
```diff
@@ -441,15 +517,31 @@ func (ps *ProxySession) doLoop(mongoConn *MongoConnectionWrapper) (*MongoConnect
...
+				}
+			}
 			ps.logTrace(ps.proxy.logger, ps.proxy.config.TraceConnPool, "failed to get a new connection. err=%v", err)
 			return nil, NewStackErrorf("cannot get connection to mongo %v using connection mode=%v readpref=%v", err, ps.proxy.config.ConnectionMode, rp)
 		}
 		ps.logTrace(ps.proxy.logger, ps.proxy.config.TraceConnPool, "got new connection %v using connection mode=%v readpref=%v", mongoConn.conn.ID(), ps.proxy.config.ConnectionMode, rp)
 	}
 
+	if ps.isMetricsEnabled {
+		requestDurationHook.StopTimer()
```
Comment (TomerYakir) [inline proxy.go:533]: [q] what would happen if we error on line 449 when starting the timer? could that potentially panic?

Comment (varunravi98) [reply proxy.go:533]: You're right, that could result in a panic. I could simply set a flag that indicates whether or not `requestDurationHook`'s timer was actually started and check that to stop the timer. The same solution could analogously be applied to `responseDurationHook`'s timer.
Alternatively, I could modify the implementation of `StopTimer()` to return an error if the underlying timer is `nil` (it shouldn't be if `StartTimer()` executed without error, and it will be if an error is returned or if `StartTimer()` wasn't called at all previously). That way, we don't have to add any checks here, but we would enforce that all types implementing `StopTimer()` must check that the underlying timer is non-nil and return an error if it is (or if anything else goes wrong when stopping the timer).

Comment (varunravi98) [reply proxy.go:536]: Per offline discussion, the first option was selected. The two flags are `isRequestTimerStarted` and `isResponseTimerStarted`, respectively.

**Review by TomerYakir — APPROVED** [review]
LGTM! 👍
